pyfr/plugins/integrate.py

```diff
@@ -0,0 +1,156 @@
+            
+            # Weights
+            rname = self.cfg.get('solver-elements-' + ename, 'soln-pts')
+            wts = get_quadrule(ename, rname, eles.nupts).wts
+            
+            # Save
+            self.elminfo.append((plocupts, rjacs, wts))
+        
+
+    def _init_gradients(self, intg):
+        # Determine what gradients, if any, are required
+        self._gradpnames = gradpnames = set()
+        for k, ex in self.exprs:
+            gradpnames.update(re.findall(r'\bgrad_(.+?)_[xyz]\b', ex))
+
+        # If gradients are required then form the relevant operators
+        if gradpnames:
+            self._gradop, self._rcpjact = [], []
+
+            for eles in intg.system.ele_map.values():
+                self._gradop.append(eles.basis.m4)
+
+                # Get the smats at the solution points
+                smat = eles.smat_at_np('upts').transpose(2, 0, 1, 3)
+
+                # Get |J|^-1 at the solution points
+                rcpdjac = eles.rcpdjac_at_np('upts')
+
+                # Product to give J^-T at the solution points
+                self._rcpjact.append(smat*rcpdjac)
+
+    def _eval_exprs(self, intg):
+
+        # Get the primitive variable names
+        pnames = self.elementscls.privarmap[self.ndims]
+
+        exprs = np.zeros(len(self.exprs))
+        # Iterate over each element type in the simulation
+        for i, (soln, (plocs, rjacs, wts)) in enumerate(zip(intg.soln, self.elminfo)):
+    
+            # Convert from conservative to primitive variables
+            psolns = self.elementscls.con_to_pri(soln.swapaxes(0, 1),
+                                                 self.cfg)
+
+            # Prepare the substitutions dictionary
+            subs = dict(zip(pnames, psolns))
+            subs.update(zip('xyz', plocs.swapaxes(0, 1)))
+
+            # Compute any required gradients
+            if self._gradpnames:
+                # Gradient operator and J^-T matrix
+                gradop, rcpjact = self._gradop[i], self._rcpjact[i]
+                nupts = gradop.shape[1]
+
+                for pname in self._gradpnames:
+                    psoln = subs[pname]
+
+                    # Compute the transformed gradient
+                    tgradpn = gradop @ psoln
+                    tgradpn = tgradpn.reshape(self.ndims, nupts, -1)
+
+                    # Untransform this to get the physical gradient
+                    gradpn = np.einsum('ijkl,jkl->ikl', rcpjact, tgradpn)
+                    gradpn = gradpn.reshape(self.ndims, nupts, -1)
+
+                    for dim, grad in zip('xyz', gradpn):
+                        subs['grad_{0}_{1}'.format(pname, dim)] = grad
+
+            for j,(k,v) in enumerate(self.exprs):
+                # Accumulate integrated evaluated expressions
+                exprs[j] += np.sum(wts[:, None]*rjacs*npeval(v, subs))
+            
+            
+        # Stack up the expressions for each element type and return
```

Comment:
Are we stacking anything here?

Reply:
agian direct copy from `tavg.py` for evaluating expressions from the ini file. Maybe a potential feature development point would be some plugin utility functions, to abstract this boiler plate stuff.

Reply:
The comment is confusing as nothing is being stacked.  In `tavg.py` the line reads:

```
# Stack up the expressions for each element type and return
return [np.dstack(exs).swapaxes(1, 2) for exs in exprs]
```
which makes sense as the expressions for each element type are kept together in an array, and thus we have a list of arrays (one for each element type).  Here we just have
```
# Stack up the expressions for each element type and return
return exprs
```
which is very confusing, as the comment mentions stacking and element types, but the code that follows is just a flat return.

Reply:
Prod comment.